File: Striver A2Z/Step 3 - Solve Problems on Arrays [Easy->Medium->Hard]/Lec 3 - Hard/merge-intervals.cpp

// https://leetcode.com/problems/merge-intervals/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans;
        int n = intervals.size();
        if(n == 0) return ans;

        sort(intervals.begin(), intervals.end());

        auto currInterval = intervals[0];
        for(int i = 1; i < n; i++){
            if(currInterval[1] >= intervals[i][0]){
                currInterval[0] = min(currInterval[0], intervals[i][0]);
                currInterval[1] = max(currInterval[1], intervals[i][1]);
            }else{
                ans.push_back(currInterval);
                currInterval = intervals[i];
            }
        }
        ans.push_back(currInterval);
        return ans;
    }
};

File: Neetcode-150-js/Intervals/merge-intervals.js

// https://leetcode.com/problems/merge-intervals/

/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    let ans = [];
    let n = intervals.length;
    if(n === 0) return ans;

    intervals = intervals.sort((a, b) => a[0] - b[0]);

    let currInterval = intervals[0];
    for(let i = 1; i < n; i++){
        if(currInterval[1] >= intervals[i][0]){
            currInterval[0] = Math.min(currInterval[0], intervals[i][0]);
            currInterval[1] = Math.max(currInterval[1], intervals[i][1]);
        }else{
            ans.push(currInterval);
            currInterval = intervals[i];
        }
    }

    ans.push(currInterval);
    return ans;
};